### ISSUE TITLE
Mypy incorrectly complains about overloads being incorrect

### ISSUE THREAD

**Issue by jhance** [python/mypy#13456]
Minimal repro:

```
from typing import overload, TypeVar, Sequence

class Entity: ...
class Assoc: ...

E = TypeVar("E", bound=Entity)
A = TypeVar("A", bound=Assoc)

class Test:
    @overload
    async def multiget_cached(
        self,
        edgecls: Sequence[E],
    ) -> None:
        ...

    @overload
    async def multiget_cached(
        self,
        edgecls: Sequence[A],
    ) -> None:
        ...

    async def multiget_cached(
        self,
        edgecls: Sequence[E] | Sequence[A],
    ) -> None:                                    
        ...
```

According to Ivan this is likely a bug in the type unifier.

**Comment by ilevkivskyi** [comment]
LOL, I just started writing an issue about this as well. It seems to me the root cause is
```python
from typing import Callable, TypeVar, Union, Sequence

TI = TypeVar("TI", bound=int)
TS = TypeVar("TS", bound=str)

f: Callable[[Sequence[TI]], None]
g: Callable[[Union[Sequence[TI], Sequence[TS]]], None]
f = g  # Bogus error: Incompatible types in assignment
```
And IMO a proper fix would be to filter out "trivially unsatisfiable" (because of upper bounds) constraints _and_ then turning `[]` into `None` before comparing constraints in `any_constraints()` (used when handling unions).

**Comment by ilevkivskyi** [comment]
Changed topic as it is not specific to overloads.

**Comment by hauntsaninja** [comment]
Another similar bug: https://github.com/python/mypy/issues/13220

**Comment by ilevkivskyi** [comment]
Btw although this is a relatively fragile part of the code, it is also a simple PR, I can try it and see what `mypy_primer` will say.

**Comment by ilevkivskyi** [comment]
@hauntsaninja That one is a bit different, where we need to select upper bound as a solution if it is "better" that the "best" constraint. IMO this is more risky to cause unpredictable fallout (but FWIW probably also an easy fix by itself).

**Comment by ilevkivskyi** [comment]
Oh, I forgot that constraints have `TypeVarId`s not `TypeVarType`, so actually it will be a significant refactoring to pass through the information about upper bounds.

Alternatively, we can just never add unsatisfiable constraints in the leaf case. But this will probably also require `eager=True` for both union cases, which is a big change (and is not very principled).

**Comment by ilevkivskyi** [comment]
Yeah, this last hack is not an option. I just tried, it works, but errors are worse, now we get something about `<nothing>` instead of e.g. `Value of type variable "T" of "twice" cannot be "int"`

**Comment by ilevkivskyi** [comment]
After all I actually decided to do the refactoring and it turned out to not be so big after all, see https://github.com/python/mypy/pull/13467
